SimulationAgent/Simulation/DeviceStatusLogic/SendTelemetry.cs

```diff
@@ -37,23 +50,87 @@ public void Setup(string deviceId, DeviceModel deviceModel)
 
             this.setupDone = true;
             this.deviceId = deviceId;
+            this.context = context;
+
+            foreach (var message in deviceModel.Telemetry)
+            {
+                this.log.Debug("Preparing telemetry timer", () =>
+                    new { this.deviceId, message.Interval.TotalSeconds, message.MessageSchema.Name, message.MessageTemplate });
+
+                var timer = this.factory.Resolve<ITimer>();
+
+                var messageContext = new SendTelemetryContext
+                {
+                    DeviceActor = context,
+                    Message = message,
+                    MessageTimer = timer,
+                    Interval = message.Interval
+                };
+
+                timer.Setup(this.Run, messageContext);
+
+                this.timers.Add(timer);
+            }
+        }
+
+        public void Start()
+        {
+            this.log.Info("Starting SendTelemetry", () => new { this.deviceId });
+            foreach (var timer in this.timers)
+            {
+                timer.RunOnce(0);
+            }
+        }
+
+        public void Stop()
+        {
+            this.log.Info("Stopping SendTelemetry", () => new { this.deviceId });
+            foreach (var timer in this.timers)
+            {
+                timer.Cancel();
+            }
         }
 
         public void Run(object context)
         {
+            var start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+
+            // Each message context contains references to device
+            // actor and the message timer, that we use to pause here
+            var messageContext = (SendTelemetryContext) context;
+            try
+            {
+                try
+                {
+                    this.RunInternalAsync(context).Wait();
+                }
+                finally
+                {
+                    var passed = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - start;
+                    messageContext?.MessageTimer?.RunOnce(messageContext.Interval.TotalMilliseconds - passed);
+                }
+            }
+            catch (ObjectDisposedException e)
+            {
+                this.log.Debug("The simulation was stopped and some of the context is not available", () => new { e });
+            }
+        }
+
+        private async Task RunInternalAsync(object context)
+        {
             this.ValidateSetup();
 
-            var callContext = (SendTelemetryContext) context;
-            var actor = callContext.Self;
-            var message = callContext.Message;
+            var messageContext = (SendTelemetryContext) context;
+            var actor = messageContext.DeviceActor;
+            var message = messageContext.Message;
 
             if (actor.CancellationToken.IsCancellationRequested)
             {
                 actor.Stop();
```

Comment:
do you need to call Stop here?  It gets stopped from the check against cosmos IIRC so you shouldn't need to; i.e. if you just returned you'd be fine (& potentially avoid errors upon shutdown).  Not critical.

Reply:
you might be right, I'd like to review the start/stop behavior eventually, possibly when we work on add/remove devices by ID. For the time being I'll leave it here